crates/iceberg/src/writer/file_writer/parquet_writer.rs

```diff
@@ -0,0 +1,390 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! The module contains the file writer for parquet file format.
+
+use std::{
+    collections::HashMap,
+    sync::{atomic::AtomicI64, Arc},
+};
+
+use crate::Result;
+use crate::{
+    io::{FileIO, OutputFile},
+    spec::{DataFileBuilder, DataFileFormat},
+    writer::CurrentFileStatus,
+    Error,
+};
+use arrow_schema::SchemaRef;
+use parquet::{arrow::AsyncArrowWriter, format::FileMetaData};
+use parquet::{arrow::PARQUET_FIELD_ID_META_KEY, file::properties::WriterProperties};
+
+use super::{
+    location_generator::LocationGenerator, track_writer::TrackWriter, FileWriter, FileWriterBuilder,
+};
+
+/// ParquetWriterBuilder is used to builder a [`ParquetWriter`]
+#[derive(Clone)]
+pub struct ParquetWriterBuilder {
+    /// `buffer_size` determines the initial size of the intermediate buffer.
+    /// The intermediate buffer will automatically be resized if necessary
+    init_buffer_size: usize,
+    props: WriterProperties,
+    schema: SchemaRef,
+
+    file_io: FileIO,
+    location_generator: Arc<dyn LocationGenerator>,
```

Comment:
Is it possible to make `ParquetWriterBuilder` generics on `LocationGenerator`?

Reply:
I can try. But what's the benefit of making LocationGenerator generic? Seems the `LocationGenerator` will not be called frequently.🤔

Reply:
The reason is mainly about performance. Usually generics have better performance, but it may make it difficult to use for user. In this case I think it's reasonable to use generics, because it's not used by user directly.